crates/starfish/core/src/block_header.rs

```diff
@@ -417,13 +417,47 @@ impl AsRef<[u8]> for BlockHeaderDigest {
 // explicitly the transaction data.
 #[derive(Clone, Copy, Serialize, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord)]
 pub struct TransactionsCommitment([u8; starfish_config::DIGEST_LENGTH]);
+pub type MerkleProofBytes = Vec<u8>;
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub(crate) struct ShardWithProof {
+    pub(crate) shard: Shard,
+    pub(crate) transaction_commitment: TransactionsCommitment,
+    pub(crate) proof: MerkleProofBytes,
+    pub(crate) block_ref: BlockRef,
+}
+
+impl PartialEq for ShardWithProof {
+    fn eq(&self, other: &Self) -> bool {
+        self.transaction_commitment == other.transaction_commitment
```

Comment:
ShardWithProof is a single shard of many for a TransactionCommitment, so 2 different shards of the same transaction_commitment should not be compared through the transaction_commitment field.